File: solutions/kotlin/dnd-character/4/src/main/kotlin/DndCharacter.kt
import kotlin.random.nextInt
import kotlin.random.Random

class DndCharacter {
    val strength: Int = ability()
    val dexterity: Int = ability()
    val constitution: Int = ability()
    val intelligence: Int = ability()
    val wisdom: Int = ability()
    val charisma: Int = ability()
    val hitpoints: Int = modifier(constitution) + 10

    companion object {
        fun ability(): Int =
            generateSequence{
                Random.nextInt(1..6)
            }.take(4).sortedDescending().take(3).sum()

        fun modifier(score: Int): Int =
            (score - 10).floorDiv(2)
    }
}

File: solutions/go/dnd-character/4/dnd_character.go
package dndcharacter

import "math/rand/v2"

type Character struct {
	Strength, Dexterity, Constitution,
    Intelligence, Wisdom, Charisma, Hitpoints int
}

func Modifier(score int) int {
    if (score - 10) % 2 < 0 {
        return (score - 10) / 2 - 1
    }
	return (score - 10) / 2
}

func Ability() (sum int) {
	min := DiceRoll()
    for i := 0; i < 3; i++ {
        d := DiceRoll()
        if min <= d {
            sum += d
        } else {
            min, sum = d, sum + min
        }
    }
    return
}

func Abilities() (str, dex, con, itl, wis, chr int) {
    return Ability(), Ability(), Ability(),
    	Ability(), Ability(), Ability()
}

func GenerateCharacter() Character {
    str, dex, con, itl, wis, chr := Abilities()
	return Character{
        str, dex, con, itl,
        wis, chr, 10 + Modifier(con),
    }
}

func DiceRoll() int {
    return rand.N(6) + 1
}